### ISSUE TITLE
Connections lasting in CLOSE_WAIT indefinitely...

### ISSUE THREAD

**Issue by fusionfoto** [haproxy/haproxy#1179]
## Detailed description of the problem

We are seeing thousands of connections left in CLOSE_WAIT. Not all of them, but some are left open for 4 or more hours with no bytes pending in either direction. We've just begun gathering metrics on the magnitude of the problem. We compare the IP in the CLOSE_WAIT table to the last known access in the haproxy.log to deduce the length of survival.

On a server with about 200 steady-state ESTABLISHED connections, we have between 1K and 2K+ in CLOSE_WAIT at any one time. Just started graphing this, so will have more data in the next few days. Only about 17 are in the TIME_WAIT status at any one time... 

## Expected behavior

We'd normally expect haproxy to close the connection within minutes of the final packet flows. And within an order of magnitude of TIME_WAIT.

## Steps to reproduce the behavior

Not sure how to reliably reproduce this at this time -- or rather, this set up seems to produce this behavior continuously.

1. 
2. 
3. 

## Do you have any idea what may have caused this?

Testing a hunch, it seems the vast majority (200x) of connections in CLOSE_WAIT incorrectly are HTTP/2.0 originally, but since we accept HTTP/2, I don't know if that is a function of popularity of HTTP/2 or not.

## What is your configuration?

```
./sanitizeconfig.sh
global
        maxconn         100000
        log /dev/log    local0
        chroot /var/lib/haproxy
        stats socket /run/haproxy/admin.sock mode 777 level admin expose-fd listeners
        stats timeout 30s
        user haproxy
        group haproxy
        daemon
        nbproc 1
        nbthread 6
        hard-stop-after 24h


        tune.h2.initial-window-size 4096000

defaults
        log     global
        mode    http
        option forwardfor
        option redispatch
        option log-separate-errors

        timeout client 8h
        timeout server 8h
        timeout connect 60s

        log-format "%ci:%cp [%tr] %ft %b/%s %TR/%Tw/%Tc/%Tr/%Ta %ST %B %CC %CS %tsc %ac/%fc/%bc/%sc/%rc %sq/%bq %hr %hs %{+Q}r %sslc"

        errorfile 400 /etc/haproxy/errors/400.http
        errorfile 403 /etc/haproxy/errors/403.http
        errorfile 408 /etc/haproxy/errors/408.http
        errorfile 500 /etc/haproxy/errors/500.http
        errorfile 502 /etc/haproxy/errors/502.http
        errorfile 503 /etc/haproxy/errors/503.http
        errorfile 504 /etc/haproxy/errors/504.http

listen statsend
        bind 127.0.0.1:9000
        mode http
        stats enable
        stats hide-version
        stats scope .
        stats realm Haproxy\ Statistics
        stats uri /haproxy-stats?stats


cache objects
        total-max-size 1024
        max-object-size 2560000
        max-age 86400

frontend cp.xxx.com
     bind *:80
     bind *:443 ssl crt /etc/apache2/sites-available/xxx.com/ssl/le/ssl-certs.pem ssl-min-ver TLSv1.2  alpn h2,h2c,http/1.1
     maxconn 50000
     compression algo gzip
     compression type text/html text/plain text/javascript application/javascript application/xml text/css
     option forwardfor
     option http-keep-alive
     timeout client     8h
     timeout http-keep-alive 60s
     timeout http-request 60s
     timeout client-fin 60s
     http-request cache-use objects
     http-response cache-store objects
     http-request set-header X-Forwarded-Port %[dst_port]
     http-request add-header X-Forwarded-Proto https if { ssl_fc }
     option http-server-close
     capture request header Referrer len 64
     capture request header Content-Length len 10
     capture request header User-Agent len 64
     http-request add-header  Strict-Transport-Security  max-age=15768000
     http-request redirect scheme https unless { ssl_fc }
     default_backend nodes


backend nodes
    mode http
    hash-type consistent
    option redispatch
    fullconn 40000


    timeout check 3000
    option httpchk GET /index.php
    http-check expect status 200




    retry-on empty-response conn-failure

    option log-health-checks

    balance leastconn

    cookie WSRVID insert indirect nocache maxidle 30m maxlife 24h


    server www3.xxx.com 127.0.0.5:443 check check-ssl ssl force-tlsv13 verify none sni ssl_fc_sni allow-0rtt alpn h2 cookie s3 maxconn 10000 check-alpn http/1.1
    server www4.xxx.com 127.0.0.6:443 check check-ssl ssl force-tlsv13 verify none sni ssl_fc_sni allow-0rtt alpn h2 cookie s4 maxconn 10000 check-alpn http/1.1
    server www5.xxx.com 127.0.0.7:443 check check-ssl ssl force-tlsv13 verify none sni ssl_fc_sni allow-0rtt alpn h2 cookie s5 maxconn 10000 check-alpn http/1.1
    server www6.xxx.com 127.0.0.8:443 check check-ssl ssl force-tlsv13 verify none sni ssl_fc_sni allow-0rtt alpn h2 cookie s6 maxconn 10000 check-alpn http/1.1


```

```
(paste your output here)
```

## Output of `haproxy -vv` and `uname -a`

```
haproxy -vvv
HA-Proxy version 2.2.8-7bf78d7 2021/01/13 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2025.
Known bugs: http://www.haproxy.org/bugs/bugs-2.2.8.html
Running on: Linux 4.15.0-130-generic #134-Ubuntu SMP Tue Jan 5 20:46:26 UTC 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-stringop-overflow -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1
  DEBUG   =

Feature list : +EPOLL -KQUEUE +NETFILTER +PCRE -PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=2).
Built with OpenSSL version : OpenSSL 1.1.1  11 Sep 2018
Running on OpenSSL version : OpenSSL 1.1.1  11 Sep 2018
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.39 2016-06-14
Running on PCRE version : 8.39 2016-06-14
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 7.5.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
              h2 : mode=HTTP       side=FE|BE     mux=H2
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : none

Available filters :
        [SPOE] spoe
        [COMP] compression
        [TRACE] trace
        [CACHE] cache
        [FCGI] fcgi-app



```

## If HAProxy crashed: Last outputs and backtraces

```
(paste your output here)
```

## Additional information (if helpful)

will update this if I generate any more data. We are also tracking issues with haproxy failing all of the backends af about 5 days of uptime (2s up, 2s down for each) and hanging outright after a few hours when we updated to 2.2.10... so there are some code changes impacting a number of behaviors.

**Comment by wtarreau** [comment]
There were known issues resulting from a fix in 2.2.8 causing CLOSE_WAIT in H2 that were solved in 2.2.10, and I'm pretty certain that this is what you're facing. As such I'd say that the real problem here is the issue you've been facing on 2.2.10 that is preventing you from updating, and which does not correspond to anything known. So it would be useful if you could share a bit more info about this (ideally in a new issue and we'd close this one as duplicate of the previous ones since fixed).

**Comment by fusionfoto** [comment]
This sounds reasonable! Let me find some time to sandbox a 2.2.10 with some live traffic and will create a new issue there. In the meantime, I will close this. Thank you so much for the rapid response!
